### PR TITLE
[WIP] Added compiler service memory analysis

### PR DESCRIPTION
**Pull request by TIHan**
This adds a little tool in Benchmarks.sln that will analyze memory consumption in the compiler service.

I'm not exactly sure how it's going to work yet, but wanted to make a PR to show current progress.

### CODE REVIEW

File: benchmarks/CompilerServiceMemoryAnalysisRunner/Program.fs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+let rec getFullMemorySize (instance: ClrInstance) (objSet: HashSet<uint64>) =
+    match instance with
+    | ClrInstance.Object(clrInstance) ->
+
+        if clrInstance.IsNull || not (objSet.Add clrInstance.Address) then 0
+        else
+
+        let instanceFieldFullMemorySize =
+            clrInstance.Type.Fields
```
Comment (dsyme) [inline benchmarks/CompilerServiceMemoryAnalysisRunner/Program.fs]: Can be
```fsharp
            clrInstance.Type.Fields
            |> Seq.sumBy (fun clrInstanceField ->
                if clrInstanceField.Type.IsValueClass then
                    let clrValueClass = clrInstance.GetValueClassField(clrInstanceField.Name)
                    getFullMemorySize (ClrInstance.ValueClass clrValueClass) objSet
                elif clrInstanceField.Type.IsPrimitive then
                    clrInstanceField.Type.BaseSize
                else
                    let clrObject = clrInstance.GetObjectField(clrInstanceField.Name)
                    getFullMemorySize (ClrInstance.Object clrObject) objSet
            )
```

File: benchmarks/CompilerServiceMemoryAnalysisRunner/Program.fs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            )
+            |> Seq.sumBy (fun f -> f ())
+
+        clrInstance.Type.BaseSize + instanceFieldFullMemorySize
+
+    | ClrInstance.ValueClass(clrInstance) ->
+
+        let instanceFieldFullMemorySize =
+            clrInstance.Type.Fields
+            |> Seq.map (fun clrInstanceField ->
```
Comment (dsyme) [inline benchmarks/CompilerServiceMemoryAnalysisRunner/Program.fs]: likewise here

File: benchmarks/CompilerServiceMemoryAnalysisRunner/Program.fs (hunk 1 above)
Comment (TIHan) [reply benchmarks/CompilerServiceMemoryAnalysisRunner/Program.fs]: Yup, I also need to hoist some of this out as I got duplicate code.

**Comment by TIHan** [comment]
Closing as I don't think we need this kind of memory analysis anymore. Through testing, we have determined that most of the memory cost is coming from `Entity`, `Val` and `TcSymbolUseData []` types in very large solutions. 

Perhaps we can revisit something like this later. I prototyped a wpf-built tool to do some analysis outside of this repo. Perhaps I could continue with that.

**Pull request closed without merging**